File: test/integration/ContainerViewsProbe.py
#===----------------------------------------------------------------------===#
#
# Part of the OpenCyphal project, under the MIT licence
# SPDX-License-Identifier: MIT
#
#===----------------------------------------------------------------------===#
#
# The C probe's reading of a container holding a view, in Python: the view is a memoryview of
# the buffer, and a write to the buffer shows through it.
#
#===----------------------------------------------------------------------===#

import struct, sys
sys.path.insert(0, sys.argv[1])
from dsdl_gen.fixtures_views.vendor.frame_1_0 import Frame
from dsdl_gen.fixtures_views.vendor.leading_1_0 import Leading
from dsdl_gen.fixtures_views.vendor.track_1_0 import Track
from dsdl_gen.fixtures_aliasable.vendor.pose_1_0 import Pose
from dsdl_gen.fixtures_aliasable.vendor.vec3_1_0 import Vec3
failures = 0
def check(what, ok):
    global failures
    print(f"  {what:<44} {'ok' if ok else 'FAILED'}")
    if not ok: failures += 1
wire = bytearray(41)
wire[0:4] = struct.pack("<I", 0x11223344)
wire[4:28] = struct.pack("<6f", 1.5, -2.25, 3.0, 4.0, 5.5, 6.75)
wire[28:40] = struct.pack("<3f", 7.0, 8.0, 9.0)
wire[40] = 0x5A
frame = Frame.deserialize(memoryview(wire))
check("sequence decoded", frame.sequence == 0x11223344)
check("view points into the buffer", frame.pose.obj is wire and frame.pose.nbytes == 24 and bytes(frame.pose) == bytes(wire[4:28]))
wire[4] ^= 0xFF
check("view is live, not a copy", frame.pose[0] == wire[4])
wire[4] ^= 0xFF
check("orientation.y read through the view", Vec3.get_y(Pose.get_orientation(frame.pose)) == 5.5)
check("velocity decoded, status after the view", frame.velocity.y == 8.0 and frame.status == 0x5A)
check("serialise reproduces the wire", frame.serialize() == bytes(wire))
short = Frame.deserialize(memoryview(wire)[:16])
check("short view holds what was there", short.pose.nbytes == 12 and bytes(short.pose) == bytes(wire[4:16]))
o = Pose.get_orientation(short.pose)
check("missing orientation reads as zero", len(o) == 0 and Vec3.get_y(o) == 0 and short.status == 0)
short_out = short.serialize()
check("short view serialises zero-filled", short_out[4:16] == bytes(wire[4:16]) and short_out[16:28] == bytes(12))
fresh = Frame()
check("fresh object holds an empty view", fresh.pose.nbytes == 0)
check("empty view serialises as zeros", fresh.serialize()[4:28] == bytes(24))
track_wire = bytearray(99)
track_wire[0] = 0x07
for at, pose in zip((1, 25, 50, 74), ((1, 2, 3, 4, 5, 6), (10, 20, 30, 40, 50, 60), (-1, -2, -3, -4, -5, -6), (0.5, 1.5, 2.5, 3.5, 4.5, 5.5))):
    track_wire[at:at + 24] = struct.pack("<6f", *pose)
track_wire[49] = 2
track_wire[98] = 0x3C
track = Track.deserialize(memoryview(track_wire))
check("pair elements are views into the buffer", all(p.obj is track_wire for p in track.pair) and bytes(track.pair[0]) == bytes(track_wire[1:25]) and bytes(track.pair[1]) == bytes(track_wire[25:49]))
check("trail keeps its count, elements are views", len(track.trail) == 2 and track.trail[1].obj is track_wire and bytes(track.trail[1]) == bytes(track_wire[74:98]))
track_wire[25] ^= 0xFF
check("element view is live, not a copy", track.pair[1][0] == track_wire[25])
track_wire[25] ^= 0xFF
check("orientation.y read through pair[1]", Vec3.get_y(Pose.get_orientation(track.pair[1])) == 50.0 and track.kind == 0x07 and track.status == 0x3C)
check("track serialise reproduces the wire", track.serialize() == bytes(track_wire))
short_track = Track.deserialize(memoryview(track_wire)[:37])
check("short track: pair[1] short, trail empty", short_track.pair[1].nbytes == 12 and bytes(short_track.pair[1]) == bytes(track_wire[25:37]) and len(short_track.trail) == 0 and short_track.status == 0)
short_track_out = short_track.serialize()
check("short element serialises zero-filled", len(short_track_out) == 51 and short_track_out[25:37] == bytes(track_wire[25:37]) and short_track_out[37:51] == bytes(14))
fresh_track = Track()
check("fresh object holds empty element views", len(fresh_track.pair) == 2 and all(p.nbytes == 0 for p in fresh_track.pair) and len(fresh_track.trail) == 0)
check("empty element views serialise as zeros", fresh_track.serialize() == bytes(51))
lead_wire = bytearray(25)
lead_wire[0:24] = struct.pack("<6f", 0.25, -0.5, 0.75, -1.0, 1.25, -1.5)
lead_wire[24] = 0xA5
lead = Leading.deserialize(memoryview(lead_wire))
check("leading view is the buffer itself", lead.pose.obj is lead_wire and lead.pose.nbytes == 24 and bytes(lead.pose) == bytes(lead_wire[:24]))
lead_wire[0] ^= 0xFF
check("leading view is live, not a copy", lead.pose[0] == lead_wire[0])
lead_wire[0] ^= 0xFF
check("orientation.y read through the leading view", Vec3.get_y(Pose.get_orientation(lead.pose)) == 1.25 and lead.status == 0xA5)
check("leading serialise reproduces the wire", lead.serialize() == bytes(lead_wire))
short_lead = Leading.deserialize(memoryview(lead_wire)[:12])
check("short leading view holds what was there", short_lead.pose.obj is lead_wire and short_lead.pose.nbytes == 12 and short_lead.status == 0)
short_lead_out = short_lead.serialize()
check("short leading view serialises zero-filled", len(short_lead_out) == 25 and short_lead_out[:12] == bytes(lead_wire[:12]) and short_lead_out[12:] == bytes(13))
empty_lead = Leading.deserialize(memoryview(b""))
o = Pose.get_orientation(empty_lead.pose)
check("empty buffer leaves an empty leading view", empty_lead.pose.nbytes == 0 and empty_lead.status == 0 and len(o) == 0 and Vec3.get_y(o) == 0)
fresh_lead = Leading()
check("fresh object holds an empty leading view", fresh_lead.pose.nbytes == 0)
check("empty leading view serialises as zeros", fresh_lead.serialize() == bytes(25))
print(f"container-views Python: {'ok' if failures == 0 else 'FAILED'}")
sys.exit(0 if failures == 0 else 1)

File: test/integration/ContainerViewsProbe.ts
//===----------------------------------------------------------------------===//
//
// Part of the OpenCyphal project, under the MIT licence
// SPDX-License-Identifier: MIT
//
//===----------------------------------------------------------------------===//
//
// The C probe's reading of a container holding a view, in TypeScript: the view is a subarray of the buffer. Compiled under noUnusedLocals.
//
//===----------------------------------------------------------------------===//

import * as frameModule from "./fixtures_views/vendor/frame_1_0";
import * as leadingModule from "./fixtures_views/vendor/leading_1_0";
import * as trackModule from "./fixtures_views/vendor/track_1_0";
import * as poseModule from "./fixtures_aliasable/vendor/pose_1_0";
import * as vec3Module from "./fixtures_aliasable/vendor/vec3_1_0";
let failures = 0;
const check = (what: string, ok: boolean): void => { console.log(`  ${what.padEnd(44)} ${ok ? "ok" : "FAILED"}`); if (!ok) failures += 1; };
const wire = new Uint8Array(41);
const view = new DataView(wire.buffer);
view.setUint32(0, 0x11223344, true);
[1.5, -2.25, 3.0, 4.0, 5.5, 6.75].forEach((v, i) => view.setFloat32(4 + i * 4, v, true));
[7.0, 8.0, 9.0].forEach((v, i) => view.setFloat32(28 + i * 4, v, true));
wire[40] = 0x5a;
const decoded = frameModule.deserializeFrame(wire);
const frame = decoded.value;
check("deserialise accepted", decoded.consumed === 41);
check("sequence decoded", frame.sequence === 0x11223344);
check("view points into the buffer", frame.pose.buffer === wire.buffer && frame.pose.byteOffset === 4 && frame.pose.length === 24);
check("orientation.y read through the view", vec3Module.getVec3Y(poseModule.getPoseOrientation(frame.pose)) === 5.5);
check("velocity decoded, status after the view", frame.velocity.y === 8.0 && frame.status === 0x5a);
const out = frameModule.serializeFrame(frame);
check("serialise reproduces the wire", out.length === 41 && out.every((b, i) => b === wire[i]));
const short = frameModule.deserializeFrame(wire.subarray(0, 16)).value;
check("short view holds what was there", short.pose.length === 12 && short.pose.byteOffset === 4);
const o = poseModule.getPoseOrientation(short.pose);
check("missing orientation reads as zero", o.length === 0 && vec3Module.getVec3Y(o) === 0 && short.status === 0);
const shortOut = frameModule.serializeFrame(short);
check("short view serialises zero-filled", shortOut.subarray(4, 16).every((b, i) => b === wire[4 + i]) && shortOut.subarray(16, 28).every((b) => b === 0));
const fresh = frameModule.makeFrame();
check("fresh object holds an empty view", fresh.pose.length === 0);
const freshOut = frameModule.serializeFrame(fresh);
check("empty view serialises as zeros", freshOut.subarray(4, 28).every((b) => b === 0));
const trackWire = new Uint8Array(99);
const trackView = new DataView(trackWire.buffer);
trackWire[0] = 0x07;
[[1, 2, 3, 4, 5, 6], [10, 20, 30, 40, 50, 60], [-1, -2, -3, -4, -5, -6], [0.5, 1.5, 2.5, 3.5, 4.5, 5.5]].forEach((pose, p) => pose.forEach((v, i) => trackView.setFloat32([1, 25, 50, 74][p] + i * 4, v, true)));
trackWire[49] = 2;
trackWire[98] = 0x3c;
const trackDecoded = trackModule.deserializeTrack(trackWire);
const track = trackDecoded.value;
check("track deserialise accepted", trackDecoded.consumed === 99);
check("pair elements are views into the buffer", track.pair[0].buffer === trackWire.buffer && track.pair[0].byteOffset === 1 && track.pair[0].length === 24 && track.pair[1].byteOffset === 25 && track.pair[1].length === 24);
check("trail keeps its count, elements are views", track.trail.length === 2 && track.trail[1].byteOffset === 74 && track.trail[1].length === 24);
check("orientation.y read through pair[1]", vec3Module.getVec3Y(poseModule.getPoseOrientation(track.pair[1])) === 50.0 && track.kind === 0x07 && track.status === 0x3c);
const trackOut = trackModule.serializeTrack(track);
check("track serialise reproduces the wire", trackOut.length === 99 && trackOut.every((b, i) => b === trackWire[i]));
const shortTrack = trackModule.deserializeTrack(trackWire.subarray(0, 37)).value;
check("short track: pair[1] short, trail empty", shortTrack.pair[1].byteOffset === 25 && shortTrack.pair[1].length === 12 && shortTrack.trail.length === 0 && shortTrack.status === 0);
const shortTrackOut = trackModule.serializeTrack(shortTrack);
check("short element serialises zero-filled", shortTrackOut.length === 51 && shortTrackOut.subarray(25, 37).every((b, i) => b === trackWire[25 + i]) && shortTrackOut.subarray(37, 51).every((b) => b === 0));
const freshTrack = trackModule.makeTrack();
check("fresh object holds empty element views", freshTrack.pair.length === 2 && freshTrack.pair.every((p) => p.length === 0) && freshTrack.trail.length === 0);
const freshTrackOut = trackModule.serializeTrack(freshTrack);
check("empty element views serialise as zeros", freshTrackOut.length === 51 && freshTrackOut.subarray(1, 51).every((b) => b === 0));
const leadWire = new Uint8Array(25);
const leadView = new DataView(leadWire.buffer);
[0.25, -0.5, 0.75, -1.0, 1.25, -1.5].forEach((v, i) => leadView.setFloat32(i * 4, v, true));
leadWire[24] = 0xa5;
const leadDecoded = leadingModule.deserializeLeading(leadWire);
const lead = leadDecoded.value;
check("leading deserialise accepted", leadDecoded.consumed === 25);
check("leading view is the buffer itself", lead.pose.buffer === leadWire.buffer && lead.pose.byteOffset === 0 && lead.pose.length === 24);
check("orientation.y read through the leading view", vec3Module.getVec3Y(poseModule.getPoseOrientation(lead.pose)) === 1.25 && lead.status === 0xa5);
const leadOut = new Uint8Array(64).fill(0xee);
check("leading serialise reproduces the wire", leadingModule.serializeLeadingInto(lead, leadOut) === 25 && leadOut.subarray(0, 25).every((b, i) => b === leadWire[i]));
const shortLead = leadingModule.deserializeLeading(leadWire.subarray(0, 12)).value;
check("short leading view holds what was there", shortLead.pose.buffer === leadWire.buffer && shortLead.pose.byteOffset === 0 && shortLead.pose.length === 12 && shortLead.status === 0);
const shortLeadOut = new Uint8Array(64).fill(0xee);
check("short leading view serialises zero-filled", leadingModule.serializeLeadingInto(shortLead, shortLeadOut) === 25 && shortLeadOut.subarray(0, 12).every((b, i) => b === leadWire[i]) && shortLeadOut.subarray(12, 25).every((b) => b === 0));
const emptyDecoded = leadingModule.deserializeLeading(new Uint8Array(0));
const emptyOrientation = poseModule.getPoseOrientation(emptyDecoded.value.pose);
check("empty buffer leaves an empty leading view", emptyDecoded.consumed === 0 && emptyDecoded.value.pose.length === 0 && emptyDecoded.value.status === 0 && emptyOrientation.length === 0 && vec3Module.getVec3Y(emptyOrientation) === 0);
const freshLead = leadingModule.makeLeading();
check("fresh object holds an empty leading view", freshLead.pose.length === 0);
const freshLeadOut = new Uint8Array(64).fill(0xee);
check("empty leading view serialises as zeros", leadingModule.serializeLeadingInto(freshLead, freshLeadOut) === 25 && freshLeadOut.subarray(0, 25).every((b) => b === 0));
console.log(`container-views TypeScript: ${failures === 0 ? "ok" : "FAILED"}`);
if (failures !== 0) { throw new Error("container-views TypeScript probe failed"); }
